File: arkUtil/arkUtil.py

# Standard modules
import os
import sys
import re
import time
import datetime
import random
import hashlib
import commentjson
import types
from StringIO import StringIO
import traceback

# only need to do this once
random.seed(time.time())

def sort(data):
	convert_lambda = lambda text: int(text) if text.isdigit() else text
	natural_sort_lambda = lambda text: [convert_lambda(text_fragment) for text_fragment in
			 re.split("([0-9]+)", text)]
	return sorted(data, key=natural_sort_lambda)

def pad(num, padding, padChar='0'):
	'''
	Pads a number, <num> with zeros so the resulting string is <padding> digits long.
	'''
	num = str(num)
	if len(num) >= padding:
		return num
	return padChar * (padding - len(num)) + num

def clamp(num, mininum, maximum):
	'''
	given a num, a min, and a max, ensures an output within that range.
	'''
	return min(max(num, mininum), maximum)

def varType(val):
	'''
	Returns variable type of value passed in.
	'''
	typeString = str(type(val))
	matches = re.findall("'([a-zA-Z_]+)'", typeString)
	if len(matches):
		return matches[0]
	return typeString

def parseJSON(options, ignoreErrors=False):
	'''
	Parses given JSON if possible.  If val is a dict, return val.
	If val is a string that can't be parsed, return None.
	If val is not dictionary or string, return {}.
	'''
	if not options:
		return {}
	if varType(options) == 'dict':
		return options
	elif varType(options) == 'str':
		try:
			parsed = commentjson.loads(options)
			return unicodeToString(parsed)
		except Exception as err:
			if not ignoreErrors:
				raise err
	elif varType(options) == 'file':
		try:
			parsed = commentjson.load(options)
			return unicodeToString(parsed)
		except Exception as err:
			if not ignoreErrors:
				raise err
			return {}
	return {}

def splitFrameRangeByChunk(frameRange, numChunks):
	'''
	Splits single frame range dict into a list of frameRange dicts
	'''
	numFrames = frameRange['endFrame'] - frameRange['startFrame'] + 1
	framesPerThread = int(numFrames / numChunks)

	frameRangeChunks = []
	startFrame = frameRange['startFrame']

	for i in range(numChunks):
		endFrame = startFrame + framesPerThread - 1

		if i == numChunks - 1:
			endFrame = frameRange['endFrame']

		frameRangeDict = {'startFrame': startFrame, 'endFrame': endFrame}
		frameRangeChunks.append(frameRangeDict)
		startFrame = endFrame + 1

	return frameRangeChunks

def postString(args):
	'''
	Formats args object for a post request.
	'''
	data = ''
	for k,v in args.iteritems():
		data += '%s=%s&' % (k,str(v).replace('%','%25'))
	return data[:-1]

def mergeDict(source, destination):
	'''
	Merges the items of two dictionaries.
	'''
	# return dict(a.items() + b.items())
	for key, value in destination.items():
		if type(value) == dict:
			# get node or create one
			node = source.setdefault(key, {})
			mergeDict(node, value)
		# fix: better list merging
		# elif type(value) == list:
		# 	node = source.setdefault(key, [])
		# 	if type(node) == list:
		# 		source[key] = value + node
		# 	else:
		# 		source[key] = value
		else:
			source[key] = value

	return source


def movieSafeDim(dim):
	'''
	Rounds number DOWN to the nearest multiple of 4.
	'''
	return int(int(dim) * .25) * 4;

def safeFilename(filename):
	'''
	Removes unsafe characters from file names.
	'''
	validChars = '-_.()abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
	return ''.join(c for c in filename if c in validChars)


def utcNow():
	'''
	Returns current UTC time.
	'''
	return int(time.time() + time.timezone)

def randomHash(length=16):
	'''
	Returns random sha224 hash of given length.  Defaults to 16.
	'''
	return hashlib.sha224(str(random.random())).hexdigest()[:length]

def makeArrayUnique(val, transformFunc=None):
	'''
	Returns given array with all unique elements.
	'''
	def makeUnique(val, transformFunc):
		seen = set()
		if transformFunc is None:
			for x in val:
				if x in seen:
					continue
				seen.add(x)
				yield x
		else:
			for x in val:
				x = transformFunc(x)
				if x in seen:
					continue
				seen.add(x)
				yield x
	return list(makeUnique(val, transformFunc))

def makeWebSafe(val):
	'''
	Takes a string and converts all non-alphanumeric characters
	to underscores.	Makes all characters lowercase.
	'''
	val = str(val)
	# Everything not a letter or number becomes an underscore
	val = re.sub('[^A-Za-z0-9]', '_', val)
	# Consecutive underscores become one dash
	val = re.sub('\_+', '_', val)
	# Leading underscores go away
	val = re.sub('^\_', '', val)
	# Trailing underscores go away
	val = re.sub('\_$', '', val)
	return val

def parseCommaArray(val):
	'''
	Turns 'likes, comments' into ['like', 'comments']
	'''
	return [i.strip() for i in val.split(',')]

def parseFrameRange(frameRanges):
	'''
	Turns '1-3,5' into [1,2,3,5]
	Turns '12-11' into [11, 12]
	'''

	if not frameRanges:
		return []
	elif type(frameRanges) is float:
		return [int(frameRanges)]
	elif type(frameRanges) is int:
		return [frameRanges]

	frameRanges = frameRanges.replace(' ', '')
	frameRanges = frameRanges.split(',')
	frames = []
	for frameRange in frameRanges:
		parts = frameRange.split('-')
		if len(parts) == 2:
			start = parseInt(parts[0])
			end = parseInt(parts[1])
			if end > start:
				frames += range(start, end + 1)
			else:
				frames += range(end, start + 1)
		elif len(parts) == 1:
			frames.append(parseInt(parts[0]))

	frames = makeArrayUnique(frames)
	frames.sort()
	return frames

def appendOrSetArray(obj, val):
	'''
	Appends val to obj if obj is an array.
	If obj is not, and val is not null, returns val as an array.
	If val is null, returns [].
	'''
	if isinstance(obj, list):
		obj.append(val)
		return obj
	if isinstance(val, list):
		return val
	if val != None:
		return [val]
	return []

def ensureArray(val):
	'''
	Returns val as an array.
	If val is null, returns [].
	'''
	if (isinstance(val, list)):
		return val
	if isinstance(val, tuple) or isinstance(val, set):
		return list(val)
	if (val == None):
		return []
	return [val]

# fix: should check for NaN, etc
def ensureNumber(val):
	'''
	Returns float version of val if val is number or a string representing a number.
	Otherwise returns 0.
	'''
	try:
		return float(val)
	except:
		return 0

def omitObjectKeys(d, keysToOmit):
	'''
	Given a dictionary, returns a dictionary with all entries with keys in keysToOmit omitted.
	'''
	keysToOmit = ensureArray(keysToOmit)
	newDict = {}
	for k, v in d.iteritems():
		if k not in keysToOmit:
			newDict[k] = v
	return newDict

def collectObjectKeys(d, keysToKeep):
	'''
	Given a dictionary, returns a dictionary with all entries with keys in keysToKeep.
	'''
	keysToKeep = ensureArray(keysToKeep)
	newDict = {}
	for k, v in d.iteritems():
		if k in keysToKeep:
			newDict[k] = v
	return newDict

def parseInt(val):
	'''
	Behaves like javascripts parseInt.
	Returns 0 if not a number.
	'''
	match = re.search(r'^(\d+)[.,]?\d*?', str(val))
	if match:
		return int(match.groups()[-1])
	return 0

def capitalize(string):
	'''
	Capitalizes first character in the given string.
	'''
	if string:
		return string[0].upper() + string[1:]
	else:
		return ''

# fix: implementation should match javascript
def capitalizeWords(string):
	'''
	Capitalizes the first character in each word.
	'''
	return ' '.join([capitalize(w) for w in string.split(' ')])

def formalName(name):
	'''
	returns a capitalized version of the words
	'''
	formalName = name
	firstChar = name[0]
	if (firstChar == '_'):
		formalName = formalName[1:]

	formalName = capitalizeWords(formalName)

	if (firstChar == '_'):
		formalName = firstChar + formalName

	return formalName

def parseSort(field, order):
	'''
	Accepts sorts in form 'field:ASC', 'field:-1', or 'field','desc'
	Returns sort object.
	'''
	parts = field.split(':')
	if len(parts) > 1:
		order = parts[1]
	else:
		order = order
	order = str(order).lower().strip()[0]
	if (order == 'd' or order == '-'):
		order = 'DSC'
	else:
		order = 'ASC'
	return {
		'field': parts[0],
		'order': order,
		'combined': parts[0] + ':' + order
	}

def stringCompare(a, b):
	'''
	Returns -1 if b > a, 1 if a > b, and 0 if a == b.
	'''
	a = a.lower()
	b = b.lower()
	if (b > a):
		return -1
	elif (a > b):
		return 1
	return 0

def getAlphaNumericOnly(val):
	'''
	Removes all non-alphanumeric values from a string.
	'''
	return ''.join([i for i in val if i.isalpha() or i.isdigit()])

def removeTrailingSlash(url):
	'''
	Ensures a trailing slash is omitted
	'''
	if url[-1] == '/':
		return url[:-1]
	return url

# purposefully omit random as python has a random library


def getRandomInteger(minimum, maximum=None):
	'''
	A wrapper around Python's random - ensures parity with helpers.js
	'''
	# if we skip max, just go from 0 to min
	if maximum is None:
		maximum = minimum
		minimum = 0
	return random.randint(minimum, maximum)

def getRandomFloat(minimum, maximum=None):
	'''
	A wrapper around Python's random - ensures parity with helpers.js
	'''
	# if we skip max, just go from 0 to min
	if maximum is None:
		maximum = minimum
		minimum = 0
	return random.uniform(minimum, maximum)

def getRandomIndex(var):
	'''
	Returns a random index number for a given array
	'''
	return getRandomInteger(len(var))

def getRandomIndexValue(var):
	'''
	Returns a random index value for a given array
	'''
	return var[getRandomIndex(var)]


def executePython(code, context={}):
	'''
	Execute python code
	'''
	# fix syntax error if last line is a comment with no new line
	if not code.endswith('\n'):
		code = code + '\n'

	# compile
	try:
		scriptType = 'exec'
		if code.count('\n') == 1:
			scriptType = 'single'
		compiled = compile(code, '<string>', scriptType)
	except Exception:
		return traceback.format_exc()

	oldStdOut = sys.stdout

	# override stdout to capture exec results
	outBuffer = StringIO()
	sys.stdout = outBuffer

	try:
		context.update(globals())
		exec(compiled, context)
	except Exception:
		# remove refernces to this command and the calling scripteditor host
		# script from the traceback output
		formatted_lines = traceback.format_exc().splitlines()
		formatted_lines.pop(2)
		formatted_lines.pop(1)
		result = '\n'.join(formatted_lines)
	else:
		result = outBuffer.getvalue()

	sys.stdout = oldStdOut
	return result

def executePythonFile(scriptPath, context={}):
	'''
	Execute a python file
	'''
	try:
		with open(scriptPath) as f:
			code = f.read()
	except Exception:
		return traceback.format_exc()

	# set expected python magic variables
	context['__file__'] = scriptPath
	context['__name__'] = '__main__'
	sys.argv = [scriptPath]

	# append the dirname to path so we can require files like normal
	sys.path.append(os.path.dirname(scriptPath))
	return executePython(code, context)


def getRegexMatches(string, regex):
	return re.findall(regex, string)

def replaceAll(string, find, replace):
	return re.sub(find, replace, string)

def defaultFor(var, defaultValue):
	if var is None:
		return defaultValue
	return var

# fix: missing moveArrayItem
# def moveArrayItem(arr, from, to):

def isError(var):
	return isinstance(var, Exception)

def isSubset(test, allItems):
	subset = True
	for item in test:
		subset = subset and item in allItems
	return subset

def joinUrl(*args):
	combined = ''
	for url in args:
		if len(combined) and combined[-1] != '/':
			combined += '/'
		if url[0] == '/':
			url = url[1:]
		combined += url
	return combined

def unicodeToString(data):
	'''
	Replaces unicode with a regular string
	for a variety of data types
	'''
	inputType = type(data)
	if isinstance(data, types.StringTypes):
		return str(data)
	elif inputType == types.ListType:
		return [unicodeToString(x) for x in data]
	elif inputType == types.DictType:
		# fix: uncomment in Sublime 3
		# return {unicodeToString(x): unicodeToString(data[x]) for x in data}
		return dict([(unicodeToString(x), unicodeToString(data[x])) for x in data])
	else:
		return data

# expand takes in a string and replaces patterns like {dd}
# with information.
# specialFunction allows a specific class to augment stringTemplate's
# functionality with class-specific replacment possibilities. See server/scripts/osOperations realPath
# method for a case of this in practice.
def expand(template, replacements={}):

	bracketedComponents = getRegexMatches(template, '{.*?\}')

	if not len(bracketedComponents):
		return template

	string = template
	for component in bracketedComponents:
		replacement = component[1:-1]
		if replacement.upper() in ['DD', 'MM', 'YY', 'YYYY']:
			replacement = getDateComponent(replacement)
		elif replacement in replacements:
			replacement = replacements[replacement]
		else:
			# otherwise keep bracketed components the same
			continue

		string = string.replace(component, replacement)

	return string

def getDateComponent(pattern):
	def datePad(dateComponent):
		return pad(dateComponent, 2)

	date = datetime.datetime.now()
	pattern = pattern.upper()

	if pattern == 'DD':
		return datePad(date.day)
	elif pattern == 'MM':
		return datePad(date.month)
	elif pattern == 'YY':
		return datePad(date.year % 100)
	elif pattern == 'YYYY':
		return str(date.year)

# formats a template to be used as a regex:
# adds capture groups around brackets
# surrounds characters around pipes with brackets, ie. [A|a] (case-insensitive options)
# escapes periods
# assumes template will start at beginning and end at end
def regexFromTemplate(template):
	if not template:
		raise ValueError('template undefined')
	regex = template
	regex = re.sub('\.', '\.', regex)
	regex = re.sub('\{[\w\d%_-]*\}', '([\w\d%_-]*)', regex)
	regex = '^' + regex + '$'
	return regex

# given a template with brackets, parses a string by extracting portions that match template
# returns as data dict
# '{a}and{b}', '1234and5' -> {a: '1234', b: '5'}
def parse(template, str):
	if not template:
		return {}

	fields = re.findall('\{([\w\d%_-]*)\}', template)
	regex = regexFromTemplate(template)

	parsed = re.match(regex, str)
	if not parsed:
		raise ValueError('could not parse string ' + str + ' with template ' + template)
	parsed = parsed.groups()
	data = dict(zip(fields, parsed))
	return data

def matchesData(template, data):
	if not template:
		raise ValueError('template undefined')
	fields = re.findall('\{([\w\d%_-]*)\}', template)
	return all([k in data.keys() for k in fields])

# needs to match to the end of the text ({asdf}/ won't match test/etc)
def matchesText(template, str):
	if not template:
		raise ValueError('template undefined')
	return bool(re.match(regexFromTemplate(template), str))


# print getPadding('%04d')
# print getPadding('.$F.')
# print getPadding('.1.')
# print getPadding('#####')

File: arkUtil/arkUtil.js

// Vendor Modules
/////////////////////////
var _ = require('lodash')
var _s = require('underscore.string')
var debug = require('debug')
debug = debug('helpers')

// Our Modules
/////////////////////////
// var constants = require('../constants')


// Constants
/////////////////////////
var seed = 156
var notAlphaNumeric = new RegExp(/[^a-z0-9_]/ig)
var validIP = new RegExp(/^(([1-9]?\d|1\d\d|2[0-5][0-5]|2[0-4]\d)\.){3}([1-9]?\d|1\d\d|2[0-5][0-5]|2[0-4]\d)$/)
var localIP = new RegExp(/(^127\.)|(^10\.)|(^172\.1[6-9]\.)|(^172\.2[0-9]\.)|(^172\.3[0-1]\.)|(^192\.168\.)/)

// Main
/////////////////////////
var helpers = module.exports = {

// Variables
/////////////////////////
isClient: typeof window != 'undefined' &&
	window.document !== undefined,
isServer: typeof process != 'undefined' &&
	process.pid !== undefined,

// Methods
/////////////////////////

pad: function(num, padding, padChar)
{
	padChar = padChar || '0'
	num = num + ''
	if (num.length >= padding)
		return num
	return new Array(padding - num.length + 1).join(padChar) + num
},

clamp: function(num, min, max)
{
  return Math.min(Math.max(num, min), max)
},

/*
Method: varType

Returns variable type of value passed in.
*/
varType: function(val)
{
	return ({})
		.toString
		.call(val)
		.match(/\s([a-zA-Z]+)/)[1]
		.toLowerCase()
},


/*
Method: parseJSON

Parses given JSON if possible.  If val is a dict, return val.
If val is a string that can't be parsed, return None.
If val is not dictionary or string, return val.
*/
parseJSON: function(val)
{
	if (_.isObject(val))
	{
		return val
	}
	else if (_.isString(val))
	{
		try {
			return JSON.parse(val)
		} catch (e) {
			return null
		}
	}
	return val
},


/*
	Method: postString

	Formats args object for a post request.
*/
postString: function(args)
{
	var data = ''
	for (var propt in args)
	{
		if (args.hasOwnProperty(propt))
		{
			data = data + String(propt) + '=' + String(args[propt]) + '&'
		}
	}
	return data.slice(0, -1)
},

/*
	Method: mergeObject

	Merges the items of two objects.
*/
mergeObject: function(a, b)
{
	return _.extend(a,b)
},

/*
	Method: movieSafeDim

	Rounds number DOWN to the nearest multiple of 4.
*/
movieSafeDim: function(dim)
{
	return  parseInt(parseInt(dim, 10) * 0.25, 10) * 4
},

/*
	Method: safeFilename

	Removes unsafe characters from file names.
*/
safeFilename: function(filename)
{
	var newName = ''
	var validChars = '-_.()abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'
	for (var i = 0; i < filename.length; i += 1)
	{
		if (validChars.indexOf(filename[i]) > -1)
			newName = newName + filename[i]
	}
	return newName
},

/*
	Method: utcNow

	Return current UTC time.
*/
utcNow: function()
{
	var d = new Date()
	return d.getTime() + d.getTimezoneOffset()
},

randomHash: function(length)
{
	return Math.random().toString(36).substring(length)
},


// fix: missing randomHash


/*
	Method: makeArrayUnique

	Returns given array with all unique elements.
*/
makeArrayUnique: function(val)
{
	// indexOf returns the first occurance of a value
	// so if indexOf value isn't equal to index
	// then the value isn't unique so trash it
	return val.filter(function(value, index, self)
		{
			return self.indexOf(value) === index
		})
},

makeWebSafe: function(val)
{
	val = String(val)
	// Everything not a letter or number becomes an underscore
	val = val.replace(/[^A-Za-z0-9]/g, '_')
	// Consecutive underscores become one dash
	val = val.replace(/\_+/g, '_')
	// Leading underscores go away
	val = val.replace(/^\_/, '')
	// Trailing underscores go away
	val = val.replace(/\_$/, '')
	return val.toLowerCase()
},

/*
	Method: parseCommaArray

	Turns 'like, comments' into ['like', 'comments']
*/
parseCommaArray: function(val)
{
	// if (!_.isString(val))
	// 	throw new Error('helpers.parseCommaArray -> val must be a string, got:' + val)
	var keys = val.split(',')
	// trim each key
	return _.map(keys, function(s)
		{
			return s.trim()
		})
},
/*
	Method: parseFrameRange

	Turns '1-3,5' into [1,2,3,5]
	Turns '12-11' into [11, 12]
*/
parseFrameRange: function(frameRanges)
{
	if (!frameRanges)
		return []
	else if (_.isNumber(frameRanges))
		return [helpers.parseInt(frameRanges)]

	frameRanges = frameRanges.split(',')
	var frames = []
	var parts, start, end
	_.each(frameRanges, function(frameRange)
	{
		parts = frameRange.split('-')
		if (parts.length == 2)
		{
			start = helpers.parseInt(parts[0])
			end = helpers.parseInt(parts[1])
			if (end > start)
				frames = frames.concat(_.range(start, end + 1))
			else
				frames = frames.concat(_.range(end, start + 1))
		}
		else if (parts.length == 1)
		{
			frames.push(helpers.parseInt(parts[0]))
		}
	})

	return _.sortBy(helpers.makeArrayUnique(frames))
},
/*
	Method: appendOrSetArray

	Appends val to obj if obj is an array.
	If obj is not, and val is not null, returns val as an array.
	If val is null, returns [].
*/
appendOrSetArray: function(obj, val)
{
	if (_.isArray(obj))
		return obj.concat(val)
	else if (_.isArray(val))
		return val
	else if (!_.isUndefined(val) && !_.isNull(val))
		return [val]
	else
		return []
},
/*
	Method: ensureArray

	Returns val as an array.
	If val is null, returns [].
*/
ensureArray: function(val)
{
	if (_.isArray(val))
		return val
	if (_.isUndefined(val))
		return []
	return [val]
},
/*
	Method: ensureNumber

	Returns int version of val if val is int or a string representing a int.
	Otherwise returns 0.
*/
ensureNumber: function(val)
{
	val = parseFloat(val)
	if (isNaN(val))
		return 0
	return val
},
/*
	Method: omitObjectKeys

	Given a dictionary, returns a dictionary with all entries with keys in keysToOmit omitted.
*/
omitObjectKeys: function(object, keysToOmit)
{
	keysToOmit = helpers.ensureArray(keysToOmit)
	// the keys need to be strings or lodash isn't in to it
	keysToOmit = _.map(keysToOmit, function(key)
	{
		return String(key)
	})
	return _.pickBy(object, function(val, key)
		{
			return !_.includes(keysToOmit, String(key))
		})
},
/*
	Method: collectObjectKeys

	Given a dictionary, returns a dictionary with all entries with keys in keysToKeep.
*/
collectObjectKeys: function(object, keysToKeep)
{
	keysToKeep = helpers.ensureArray(keysToKeep)
	// the keys need to be strings or lodash isn't in to it
	keysToKeep = _.map(keysToKeep, function(key)
	{
		return String(key)
	})
	return _.pickBy(object, function(val, key)
		{
			return _.includes(keysToKeep, key)
		})
},

/*
	Method: parseInt

	Behaves like javascripts parseInt.
	Returns 0 if not a number.
*/
parseInt: function(val)
{
	return parseInt(val, 10) || 0
},


/*
	Method: capitalize

	Capitalizes first character in the given string.
*/
capitalize: function(str)
{
	if (str)
	return str.charAt(0).toUpperCase() + str.slice(1)
},
/*
	Method: capitalizeWords

	Capitalizes the first character in each word.
*/
capitalizeWords: function(str)
{
	// http://kevin.vanzonneveld.net
	// +   original by: Jonas Raoni Soares Silva (http://www.jsfromhell.com)
	// +   improved by: Waldo Malqui Silva
	// +   bugfixed by: Onno Marsman
	// +   improved by: Robin
	// +      input by: James (http://www.james-bell.co.uk/)
	// +   improved by: Kevin van Zonneveld (http://kevin.vanzonneveld.net)
	// *     example 1: capitalizeWords('kevin van  zonneveld');
	// *     returns 1: 'Kevin Van  Zonneveld'
	// *     example 2: capitalizeWords('HELLO WORLD');
	// *     returns 2: 'HELLO WORLD'
	return (str + '')
		.replace(/^([a-z\u00E0-\u00FC])|\s+([a-z\u00E0-\u00FC])/g,
			function (word)
			{
				return word.toUpperCase()
			})
},
formalName: function(name)
{
	var formalName = name
	var firstChar = name[0]
	// slice off the 1st character
	if (firstChar == '_')
		formalName = formalName.slice(1)

	formalName = _s.humanize(formalName)
	formalName = helpers.capitalizeWords(formalName)

	// add it back on
	if (firstChar == '_')
		formalName = firstChar + formalName
	return formalName
},

/*
	Method: parseSort

	Accepts sorts in form 'field:ASC', 'field:-1',
	or 'field','desc'
	Returns sort object.
*/
parseSort: function(field, order)
{
	var parts = field.split(':')

	order = parts[1] || order
	order = String(order)[0].toLowerCase().trim()[0]
	// default to ASC
	if (order == 'd' || order == '-')
		order = 'DSC'
	else
		order = 'ASC'

	return {
		field: parts[0],
		order: order,
		combined: parts[0] + ':' + order
	}
},

/*
	Method: stringCompare

	Returns -1 if b > a, 1 if a > b, and 0 if a == b.
*/
stringCompare: function(valA, valB, localeCompare)
{
	if (localeCompare)
		return valA.toLowerCase().localeCompare(valB.toLowerCase())

	valA = valA.toLowerCase()
	valB = valB.toLowerCase()
	return valA < valB ? -1: (valA > valB ? 1: 0)
},

/*
	Method: getAlphaNumericOnly

	Removes all non-alphanumeric characters from a string.
*/
getAlphaNumericOnly: function(val)
{
	return val.replace(notAlphaNumeric, '')
},

removeTrailingSlash: function(url)
{
	// remove trailing slashes
	if (url[url.length-1] == '/')
		return url.slice(0,-1)
	return url
},

// Method: random
// Returns a random float between 0 and 1.
// Slight bias towards 0 and 1.
random: function()
{
	seed += 1
	var x = Math.sin(seed) * 10000;
	return x - Math.floor(x);
},
// Method: getRandomInteger
// Returns a random integer between min and max inclusive.
getRandomInteger: function(min, max)
{
	// if we skip max, just go from 0 to min
	if (max === undefined)
	{
		max = min
		min = 0
	}
	return Math.floor(helpers.random() * (max - min + 1)) + min
},

// Method: getRandomFloat
// Returns a random float between min and max exclusive.
getRandomFloat: function(min, max)
{
	// if we skip max, just go from 0 to min
	if (max === undefined)
	{
		max = min
		min = 0
	}
	return helpers.random() * (max - min) + min
},

// Method: getRandomIndex
// Returns a random index number for a given array
getRandomIndex: function(arr)
{
	return helpers.getRandomInteger(arr.length)
},

// Method: getRandomIndexValue
// Returns a random index value for a given array
getRandomIndexValue: function(arr)
{
	return arr[helpers.getRandomIndex(arr)]
},

// purposefully skip executePython

// purposefully skip executePythonFile

// collect all the matches in a string for a given
// regex object note: /g flag must be set
// on RegExp(regex, 'g') or /regex/g
getRegexMatches: function(str, regex)
{
	var matches = []
	var match
	do {
		match = regex.exec(str)
		if (match !== null)
			matches.push(match)
	}
	while (match !== null)
	return matches
},
replaceAll: function(str, find, replace)
{
	return str.replace(new RegExp(find, 'g'), replace)
},
defaultFor: function(variable, defaultValue)
{
	if (variable === undefined)
		return defaultValue
	return variable
},
moveArrayItem: function(arr, from, to)
{
	if (from < 0)
		from = arr.length + from
	var element = arr[from]
	arr.splice(from, 1)
	arr.splice(to, 0, element)
	return arr
},

isError: function(obj)
{
	return obj instanceof Error
},

isSubset: function(test, all)
{
	var subset = true
	_.each(test, function(val)
	{
		subset = subset && _.includes(all, val)
	})
	return subset
},

/*
	Method: joinURL

	Combines all string arguments to function.
*/
joinUrl: function()
{
	if (arguments.length < 2)
		return arguments
	var combined = ''
	_.each(arguments, function(url)
	{
		if (combined.length &&
			combined.slice(-1) != '/')
			combined += '/'
		if (url[0] == '/')
			url = url.slice(1)
		combined += url
	})
	return combined
},

// Skipped unicodeToString, not needed in javascript




// Javascript only
//////////////////////////////////////////////////

// fix: should be able to make these generic
// and just do clientOnly but arguments magic variable
// doesn't seem to work reliably with tests
clientOnly: function(func, context)
{
	return function()
	{
		if (helpers.isClient)
			func.apply(context, arguments)
	}
},
serverOnly: function(func, context)
{
	return function()
	{
		if (helpers.isServer)
			func.apply(context, arguments)
	}
},

parseQueryString: function(queryString)
{
	// lob off the '?' at the head
	queryString = queryString.substring( queryString.indexOf('?') + 1 )

	var params = {}
	var queryParts = decodeURI(queryString).split(/&/g)
	_.each(queryParts, function(val)
		{
			var parts = val.split('=')
			if (parts.length >= 1)
			{
				val = undefined
				if (parts.length == 2)
					val = parts[1]
				params[parts[0]] = val
			}
		})
	return params
},

/*
	Method: getGlobal

	Gets a global given a key.
*/
getGlobal: function(key)
{
	if (!_.isUndefined(global))
		return global[key]
	else if (!_.isUndefined(window))
		return window[key]
	return undefined
},
/*
	Method: setGlobal

	Sets a global given a key/value pair.
*/
setGlobal: function(key, val)
{
	if (!_.isUndefined(global))
		global[key] = val
	else if (!_.isUndefined(window))
		window[key] = val
},

/*
	Method: isValidIP

	Tests if an IP is a valid local IP
*/
isValidIP: function(ip)
{
	// split off ipv6 for now
	ip = ip.split(':').slice(-1)
	return validIP.test(ip)
},
/*
	Method: isLocalIP

	Tests if an IP is a valid local IP
*/
isLocalIP: function(ip)
{
	if (!helpers.isValidIP(ip))
		return false

	// split off ipv6 for now
	ip = ip.split(':').slice(-1)
	return localIP.test(ip)
},

// end of module
}

helpers.jQuery = helpers.getGlobal('jQuery')

